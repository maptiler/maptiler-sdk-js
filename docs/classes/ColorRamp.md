[**MapTiler SDK v3.3.0-rc1**](../README.md)

***

[MapTiler SDK](../README.md) / ColorRamp

# Class: ColorRamp

Defined in: [src/ColorRamp.ts:64](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L64)

## Extends

- `Array`\<[`ColorStop`](../type-aliases/ColorStop.md)\>

## Indexable

\[`n`: `number`\]: [`ColorStop`](../type-aliases/ColorStop.md)

## Constructors

### Constructor

> **new ColorRamp**(`options`): `ColorRamp`

Defined in: [src/ColorRamp.ts:83](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L83)

#### Parameters

##### options

[`ColorRampOptions`](../type-aliases/ColorRampOptions.md) = `{}`

#### Returns

`ColorRamp`

#### Overrides

`Array<ColorStop>.constructor`

## Methods

### at()

> **at**(`pos`): [`ColorStop`](../type-aliases/ColorStop.md)

Defined in: [src/ColorRamp.ts:153](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L153)

Takes an integer value and returns the item at that index,
allowing for positive and negative integers.
Negative integers count back from the last item in the array.

#### Parameters

##### pos

`number`

#### Returns

[`ColorStop`](../type-aliases/ColorStop.md)

#### Overrides

`Array.at`

***

### clone()

> **clone**(): `ColorRamp`

Defined in: [src/ColorRamp.ts:161](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L161)

#### Returns

`ColorRamp`

***

### getBounds()

> **getBounds**(): `object`

Defined in: [src/ColorRamp.ts:186](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L186)

#### Returns

`object`

##### max

> **max**: `number`

##### min

> **min**: `number`

***

### getCanvasStrip()

> **getCanvasStrip**(`options`): `HTMLCanvasElement`

Defined in: [src/ColorRamp.ts:242](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L242)

#### Parameters

##### options

###### horizontal?

`boolean`

###### size?

`number`

###### smooth?

`boolean`

#### Returns

`HTMLCanvasElement`

***

### getColor()

> **getColor**(`value`, `options`): [`RgbaColor`](../type-aliases/RgbaColor.md)

Defined in: [src/ColorRamp.ts:190](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L190)

#### Parameters

##### value

`number`

##### options

###### smooth?

`boolean`

#### Returns

[`RgbaColor`](../type-aliases/RgbaColor.md)

***

### getColorHex()

> **getColorHex**(`value`, `options`): `string`

Defined in: [src/ColorRamp.ts:224](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L224)

Get the color as an hexadecimal string

#### Parameters

##### value

`number`

##### options

###### smooth?

`boolean`

###### withAlpha?

`boolean`

#### Returns

`string`

***

### getColorRelative()

> **getColorRelative**(`value`, `options`): [`RgbaColor`](../type-aliases/RgbaColor.md)

Defined in: [src/ColorRamp.ts:237](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L237)

Get the color of the color ramp at a relative position in [0, 1]

#### Parameters

##### value

`number`

##### options

###### smooth?

`boolean`

#### Returns

[`RgbaColor`](../type-aliases/RgbaColor.md)

***

### getRawColorStops()

> **getRawColorStops**(): [`ColorStop`](../type-aliases/ColorStop.md)[]

Defined in: [src/ColorRamp.ts:165](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L165)

#### Returns

[`ColorStop`](../type-aliases/ColorStop.md)[]

***

### hasTransparentStart()

> **hasTransparentStart**(): `boolean`

Defined in: [src/ColorRamp.ts:366](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L366)

Check if this color ramp has a transparent start

#### Returns

`boolean`

***

### resample()

> **resample**(`method`, `samples`): `ColorRamp`

Defined in: [src/ColorRamp.ts:283](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L283)

Apply a non-linear ressampling. This will create a new instance of ColorRamp with the same bounds.

#### Parameters

##### method

`"ease-in-square"` | `"ease-out-square"` | `"ease-in-sqrt"` | `"ease-out-sqrt"` | `"ease-in-exp"` | `"ease-out-exp"`

##### samples

`number` = `15`

#### Returns

`ColorRamp`

***

### reverse()

> **reverse**(`options`): `ColorRamp`

Defined in: [src/ColorRamp.ts:175](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L175)

Reverses the elements in an array in place.
This method mutates the array and returns a reference to the same array.

#### Parameters

##### options

###### clone?

`boolean`

#### Returns

`ColorRamp`

#### Overrides

`Array.reverse`

***

### scale()

> **scale**(`min`, `max`, `options`): `ColorRamp`

Defined in: [src/ColorRamp.ts:125](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L125)

#### Parameters

##### min

`number`

##### max

`number`

##### options

###### clone?

`boolean`

#### Returns

`ColorRamp`

***

### setStops()

> **setStops**(`stops`, `options`): `ColorRamp`

Defined in: [src/ColorRamp.ts:99](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L99)

#### Parameters

##### stops

[`ColorStop`](../type-aliases/ColorStop.md)[]

##### options

###### clone?

`boolean`

#### Returns

`ColorRamp`

***

### transparentStart()

> **transparentStart**(): `ColorRamp`

Defined in: [src/ColorRamp.ts:344](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L344)

Makes a clone of this color ramp that is fully transparant at the begining of their range

#### Returns

`ColorRamp`

***

### fromArrayDefinition()

> `static` **fromArrayDefinition**(`cr`): `ColorRamp`

Defined in: [src/ColorRamp.ts:71](https://github.com/maptiler/maptiler-sdk-js/blob/d9cb958ebf063ecde2f6f583eb172e5a83460e6a/src/ColorRamp.ts#L71)

Converts a array-definition color ramp definition into a usable ColorRamp instance.
Note: units are not converted and may need to to be converted beforehand (eg. kelvin to centigrade)

#### Parameters

##### cr

[`ArrayColorRamp`](../type-aliases/ArrayColorRamp.md)

#### Returns

`ColorRamp`
