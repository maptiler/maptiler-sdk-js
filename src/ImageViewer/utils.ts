export class FetchError extends Error {
  status: number;
  statusText: string;

  constructor(response: Response, resource: string) {
    const message = `[ImageViewer]: Failed to fetch ${resource} at ${response.url}: ${response.status.toString()}: ${response.statusText}`;

    super(message);

    this.name = "FetchError";
    this.message = message;
    this.status = response.status;
    this.statusText = response.statusText;
  }
}
