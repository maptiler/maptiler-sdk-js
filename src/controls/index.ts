export * from "./MaptilerGeolocateControl";
export * from "./MaptilerLogoControl";
export * from "./MaptilerTerrainControl";
export * from "./MaptilerNavigationControl";
export * from "./MaptilerProjectionControl";
export * from "./Minimap";
