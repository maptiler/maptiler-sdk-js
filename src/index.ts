/**
 * Maplibre export first, then extensions can overload the exports.
 */
export * from "maplibre-gl";

/**
 * To perform explicit named export so that they are included in the UMD bundle
 */
// import * as ML from "maplibre-gl";

import maplibregl from "maplibre-gl";

const {
  // supported,
  setRTLTextPlugin,
  getRTLTextPluginStatus,
  LngLat,
  LngLatBounds,
  MercatorCoordinate,
  Evented,
  AJAXError,
  prewarm,
  clearPrewarmedResources,
  version,
  workerCount,
  maxParallelImageRequests,
  workerUrl,
  addProtocol,
  removeProtocol,
} = maplibregl;

// We still want to export maplibregl.Map, but as a different name
const MapMLGL = maplibregl.Map;
const MarkerMLGL = maplibregl.Marker;
const PopupMLGL = maplibregl.Popup;
const StyleMLGL = maplibregl.Style;
const CanvasSourceMLGL = maplibregl.CanvasSource;
const GeoJSONSourceMLGL = maplibregl.GeoJSONSource;
const ImageSourceMLGL = maplibregl.ImageSource;
const RasterTileSourceMLGL = maplibregl.RasterTileSource;
const RasterDEMTileSourceMLGL = maplibregl.RasterDEMTileSource;
const VectorTileSourceMLGL = maplibregl.VectorTileSource;
const VideoSourceMLGL = maplibregl.VideoSource;
const NavigationControlMLGL = maplibregl.NavigationControl;
const GeolocateControlMLGL = maplibregl.GeolocateControl;
const AttributionControlMLGL = maplibregl.AttributionControl;
const LogoControlMLGL = maplibregl.LogoControl;
const ScaleControlMLGL = maplibregl.ScaleControl;
const FullscreenControlMLGL = maplibregl.FullscreenControl;
const TerrainControlMLGL = maplibregl.TerrainControl;

export {
  // supported,
  setRTLTextPlugin,
  getRTLTextPluginStatus,
  PopupMLGL,
  MarkerMLGL,
  StyleMLGL,
  LngLat,
  LngLatBounds,
  MercatorCoordinate,
  Evented,
  AJAXError,
  CanvasSourceMLGL,
  GeoJSONSourceMLGL,
  ImageSourceMLGL,
  RasterDEMTileSourceMLGL,
  RasterTileSourceMLGL,
  VectorTileSourceMLGL,
  VideoSourceMLGL,
  prewarm,
  clearPrewarmedResources,
  version,
  workerCount,
  maxParallelImageRequests,
  workerUrl,
  addProtocol,
  removeProtocol,
  MapMLGL,
};

// Exporting types of class instances from MapLibre:
export type NavigationControlMLGL = InstanceType<typeof NavigationControlMLGL>;
export type GeolocateControlMLGL = InstanceType<typeof GeolocateControlMLGL>;
export type AttributionControlMLGL = InstanceType<
  typeof AttributionControlMLGL
>;
export type LogoControlMLGL = InstanceType<typeof LogoControlMLGL>;
export type ScaleControlMLGL = InstanceType<typeof ScaleControlMLGL>;
export type FullscreenControlMLGL = InstanceType<typeof FullscreenControlMLGL>;
export type TerrainControlMLGL = InstanceType<typeof TerrainControlMLGL>;
export type MarkerMLGL = InstanceType<typeof MarkerMLGL>;
export type PopupMLGL = InstanceType<typeof PopupMLGL>;
export type StyleMLGL = InstanceType<typeof StyleMLGL>;
export type LngLat = InstanceType<typeof LngLat>;
export type LngLatBounds = InstanceType<typeof LngLatBounds>;
export type MercatorCoordinate = InstanceType<typeof MercatorCoordinate>;
export type Evented = InstanceType<typeof Evented>;
export type AJAXError = InstanceType<typeof AJAXError>;
export type CanvasSourceMLGL = InstanceType<typeof CanvasSourceMLGL>;
export type GeoJSONSourceMLGL = InstanceType<typeof GeoJSONSourceMLGL>;
export type ImageSourceMLGL = InstanceType<typeof ImageSourceMLGL>;
export type RasterDEMTileSourceMLGL = InstanceType<
  typeof RasterDEMTileSourceMLGL
>;
export type RasterTileSourceMLGL = InstanceType<typeof RasterTileSourceMLGL>;
export type VectorTileSourceMLGL = InstanceType<typeof VectorTileSourceMLGL>;
export type VideoSourceMLGL = InstanceType<typeof VideoSourceMLGL>;
export type MapMLGL = InstanceType<typeof MapMLGL>;

// SDK specific
import { Map, GeolocationType } from "./Map";
import type { MapOptions, LoadWithTerrainEvent } from "./Map";

import { Marker } from "./Marker";
import { Popup } from "./Popup";
import { Style } from "./Style";
import { CanvasSource } from "./CanvasSource";
import { GeoJSONSource } from "./GeoJSONSource";
import { ImageSource } from "./ImageSource";
import { RasterTileSource } from "./RasterTileSource";
import { RasterDEMTileSource } from "./RasterDEMTileSource";
import { VectorTileSource } from "./VectorTileSource";
import { VideoSource } from "./VideoSource";
import { NavigationControl } from "./NavigationControl";
import { GeolocateControl } from "./GeolocateControl";
import { AttributionControl } from "./AttributionControl";
import { LogoControl } from "./LogoControl";
import { ScaleControl } from "./ScaleControl";
import { FullscreenControl } from "./FullscreenControl";
import { TerrainControl } from "./TerrainControl";

// Import of modified versions of the controls
import { MaptilerGeolocateControl } from "./MaptilerGeolocateControl";
import { MaptilerLogoControl } from "./MaptilerLogoControl";
import { MaptilerTerrainControl } from "./MaptilerTerrainControl";
import { MaptilerNavigationControl } from "./MaptilerNavigationControl";

// importing client functions to expose them as part of the SDK
import type {
  BBox,
  Position,
  GeocodingOptions,
  CoordinatesSearchOptions,
  CenteredStaticMapOptions,
  AutomaticStaticMapOptions,
  BoundedStaticMapOptions,
} from "@maptiler/client";

import {
  geocoding,
  geolocation,
  coordinates,
  data,
  staticMaps,
  ServiceError,
  LanguageGeocoding,
  LanguageGeocodingString,
  ReferenceMapStyle,
  MapStyle,
  MapStyleVariant,
} from "@maptiler/client";

import type { MapStyleType } from "@maptiler/client";

import { Point } from "./Point";
import type { Matrix2 } from "./Point";

// Importing enums and configs
import { config, SdkConfig } from "./config";
import { Language, LanguageString, LanguageKey } from "./language";
import type { Unit } from "./unit";
import type { MinimapOptionsInput } from "./Minimap";

// Exporting types
export type {
  MapOptions,
  MinimapOptionsInput,
  LoadWithTerrainEvent,
  GeocodingOptions,
  BBox,
  Position,
  CoordinatesSearchOptions,
  CenteredStaticMapOptions,
  BoundedStaticMapOptions,
  AutomaticStaticMapOptions,
  LanguageString,
  LanguageKey,
  LanguageGeocodingString,
  Unit,
  MapStyleType,
  Matrix2,
};

// Export convert functions 'str2xml', 'xml2str', 'gpx', and 'kml'
export * from "./converters";

// Export the color ramp logic and all the built-in color ramps
export * from "./colorramp";

// Exporting classes, objects, functions, etc.
export {
  Map,
  Marker,
  Popup,
  Style,
  CanvasSource,
  GeoJSONSource,
  ImageSource,
  RasterTileSource,
  RasterDEMTileSource,
  VideoSource,
  NavigationControl,
  GeolocateControl,
  AttributionControl,
  LogoControl,
  ScaleControl,
  FullscreenControl,
  TerrainControl,
  VectorTileSource,
  GeolocationType,
  SdkConfig,
  config,
  ServiceError,
  geocoding,
  geolocation,
  coordinates,
  data,
  staticMaps,
  MapStyle,
  Language,
  LanguageGeocoding,
  Point,
  ReferenceMapStyle,
  MapStyleVariant,
  MaptilerGeolocateControl,
  MaptilerLogoControl,
  MaptilerTerrainControl,
  MaptilerNavigationControl,
};

export * from "./helpers";
