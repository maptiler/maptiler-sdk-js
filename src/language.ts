/**
 * Languages. Note that not all the languages of this list are available but the compatibility list may be expanded in the future.
 */
const Language = {
  /**
   * The visitor language mode concatenates the prefered language from the user settings and the "default name".
   * Note: The "default name" is equivalent to OSM's `{name}`, which can be the most recognized names a global
   * scale or the local name.
   * This mode is helpful in the context where a user needs to access both the local names and the names in their
   * own language, for instance when traveling abroad, where signs likely to be only available in the local language.
   */
  VISITOR: "visitor",

  /**
   * The visitor language mode concatenates English and the "default name".
   * Note: The "default name" is equivalent to OSM's `{name}`, which can be the most recognized names a global
   * scale or the local name.
   * This mode is helpful in the context where a user needs to access both the local names and the names in their
   * own language, for instance when traveling abroad, where signs likely to be only available in the local language.
   */
  VISITOR_ENGLISH: "visitor_en",

  /**
   * Language as the style is designed. Not that this is the default state and one
   * the language has been changed to another than `STYLE`, then it cannot be set back to `STYLE`.
   */
  STYLE: "style",

  /**
   * AUTO mode uses the language of the browser
   */
  AUTO: "auto",

  /**
   * STYLE is a custom flag to keep the language of the map as defined into the style.
   * If STYLE is set in the constructor, then further modification of the language
   * with `.setLanguage()` is not possible.
   */
  STYLE_LOCK: "style_lock",

  /**
   * Default fallback languages that uses latin charaters
   */
  LATIN: "name:latin",

  /**
   * Default fallback languages that uses non-latin charaters
   */
  NON_LATIN: "name:nonlatin",

  /**
   * Labels are in their local language, when available
   */
  LOCAL: "name",

  /**
   * International name
   */
  INTERNATIONAL: "name_int",

  ALBANIAN: "name:sq",
  AMHARIC: "name:am",
  ARABIC: "name:ar",
  ARMENIAN: "name:hy",
  AZERBAIJANI: "name:az",
  BASQUE: "name:eu",
  BELORUSSIAN: "name:be",
  BENGALI: "name:bn",
  BOSNIAN: "name:bs",
  BRETON: "name:br",
  BULGARIAN: "name:bg",
  CATALAN: "name:ca",
  CHINESE: "name:zh",
  TRADITIONAL_CHINESE: "name:zh-Hant",
  SIMPLIFIED_CHINESE: "name:zh-Hans",
  CORSICAN: "name:co",
  CROATIAN: "name:hr",
  CZECH: "name:cs",
  DANISH: "name:da",
  DUTCH: "name:nl",
  ENGLISH: "name:en",
  ESPERANTO: "name:eo",
  ESTONIAN: "name:et",
  FINNISH: "name:fi",
  FRENCH: "name:fr",
  FRISIAN: "name:fy",
  GEORGIAN: "name:ka",
  GERMAN: "name:de",
  GREEK: "name:el",
  HEBREW: "name:he",
  HINDI: "name:hi",
  HUNGARIAN: "name:hu",
  ICELANDIC: "name:is",
  INDONESIAN: "name:id",
  IRISH: "name:ga",
  ITALIAN: "name:it",
  JAPANESE: "name:ja",
  JAPANESE_HIRAGANA: "name:ja-Hira",
  JAPANESE_KANA: "name:ja_kana",
  JAPANESE_LATIN: "name:ja_rm",
  JAPANESE_2018: "name:ja-Latn",
  KANNADA: "name:kn",
  KAZAKH: "name:kk",
  KOREAN: "name:ko",
  KOREAN_LATIN: "name:ko-Latn",
  KURDISH: "name:ku",
  ROMAN_LATIN: "name:la",
  LATVIAN: "name:lv",
  LITHUANIAN: "name:lt",
  LUXEMBOURGISH: "name:lb",
  MACEDONIAN: "name:mk",
  MALAYALAM: "name:ml",
  MALTESE: "name:mt",
  NORWEGIAN: "name:no",
  OCCITAN: "name:oc",
  PERSIAN: "name:fa",
  POLISH: "name:pl",
  PORTUGUESE: "name:pt",
  PUNJABI: "name:pa",
  WESTERN_PUNJABI: "name:pnb",
  ROMANIAN: "name:ro",
  ROMANSH: "name:rm",
  RUSSIAN: "name:ru",
  SCOTTISH_GAELIC: "name:gd",
  SERBIAN_CYRILLIC: "name:sr",
  SERBIAN_LATIN: "name:sr-Latn",
  SLOVAK: "name:sk",
  SLOVENE: "name:sl",
  SPANISH: "name:es",
  SWEDISH: "name:sv",
  TAMIL: "name:ta",
  TELUGU: "name:te",
  THAI: "name:th",
  TURKISH: "name:tr",
  UKRAINIAN: "name:uk",
  URDU: "name:ur",
  VIETNAMIAN_LATIN: "name:vi",
  WELSH: "name:cy",
} as const;

const languagesIsoSet = new Set(Object.values(Language) as Array<string>);

function isLanguageSupported(lang: string): boolean {
  return languagesIsoSet.has(lang);
}

const languageCodeSet = new Set(Object.values(Language));

/**
 * Type representing the key of the Language object
 */
type LanguageKey = keyof typeof Language;

type Values<T> = T[keyof T];

/**
 * Built-in languages values as strings
 */
type LanguageString = Values<typeof Language>;

function getBrowserLanguage(): LanguageString {
  if (typeof navigator === "undefined") {
    return `name:${
      Intl.DateTimeFormat().resolvedOptions().locale.split("-")[0]
    }` as LanguageString;
  }

  const canditatelangs = Array.from(
    new Set(navigator.languages.map((l) => `name:${l.split("-")[0]}`)),
  ).filter((l) => languageCodeSet.has(l as LanguageString));

  return canditatelangs.length
    ? (canditatelangs[0] as LanguageString)
    : Language.LOCAL;
}

export {
  Language,
  LanguageString,
  LanguageKey,
  getBrowserLanguage,
  isLanguageSupported,
};
