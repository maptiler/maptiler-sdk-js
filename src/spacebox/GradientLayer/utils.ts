import type { LngLat } from "maplibre-gl";

function getAntipode(lngLat: LngLat) {
  const { lat, lng } = lngLat;
  const antipodeLat = -lat;

  let antipodeLng = lng + 180;

  if (antipodeLng > 180) {
    antipodeLng -= 360;
  } else if (antipodeLng < -180) {
    antipodeLng += 360;
  }

  return { lat: antipodeLat, lng: antipodeLng };
}

export { getAntipode };
