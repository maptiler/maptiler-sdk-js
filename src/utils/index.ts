export * from "./dom";
export * from "./MTAnimation";
