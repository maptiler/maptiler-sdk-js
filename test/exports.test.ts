import { describe, it, expect } from "vitest";
import packagejson from "../package.json";
import * as language from "../src/language";
import * as controls from "../src/controls";
import * as converters from "../src/converters";
import * as ColorRamp from "../src/ColorRamp";
import * as helpers from "../src/utils";

// List of expected exports
const expectedExports = Array.from(
  new Set([
    "getVersion",
    "getMapLibreVersion",
    "configMLGL",
    "getWebGLSupportError",
    "displayWebGLContextLostWarning",
    "config",
    "SdkConfig",
    "Map",
    "GeolocationType",
    "Marker",
    "Popup",
    "Style",
    "CanvasSource",
    "GeoJSONSource",
    "ImageSource",
    "RasterTileSource",
    "RasterDEMTileSource",
    "VectorTileSource",
    "VideoSource",
    "NavigationControl",
    "GeolocateControl",
    "AttributionControl",
    "LogoControl",
    "ScaleControl",
    "FullscreenControl",
    "TerrainControl",
    "BoxZoomHandler",
    "ScrollZoomHandler",
    "CooperativeGesturesHandler",
    "KeyboardHandler",
    "TwoFingersTouchPitchHandler",
    "MapWheelEvent",
    "MapTouchEvent",
    "MapMouseEvent",
    "setRTLTextPlugin",
    "getRTLTextPluginStatus",
    "LngLat",
    "LngLatBounds",
    "MercatorCoordinate",
    "Evented",
    "AJAXError",
    "prewarm",
    "clearPrewarmedResources",
    "Hash",
    "helpers",
    "Point",
    "EdgeInsets",
    "DragRotateHandler",
    "DragPanHandler",
    "TwoFingersTouchZoomRotateHandler",
    "DoubleClickZoomHandler",
    "TwoFingersTouchZoomHandler",
    "TwoFingersTouchRotateHandler",
    "getWorkerCount",
    "setWorkerCount",
    "getMaxParallelImageRequests",
    "setMaxParallelImageRequests",
    "getWorkerUrl",
    "setWorkerUrl",
    "addSourceType",
    "importScriptInWorkers",
    "addProtocol",
    "removeProtocol",
    "MapStyle",
    "MapStyleVariant",
    "ReferenceMapStyle",
    "ServiceError",
    "bufferToPixelDataBrowser",
    "circumferenceAtLatitude",
    "coordinates",
    "data",
    "elevation",
    "expandMapStyle",
    "geocoding",
    "geolocation",
    "getBufferToPixelDataParser",
    "getTileCache",
    "mapStylePresetList",
    "math",
    "misc",
    "staticMaps",
    "styleToStyle",
    "areSameLanguages",
    "toLanguageInfo",
    "isLanguageInfo",
    "getAutoLanguage",
    "getLanguageInfoFromFlag",
    "getLanguageInfoFromCode",
    "getLanguageInfoFromKey",
    "getWebGLSupportError",
    "displayWebGLContextLostWarning",
    "config",
    "SdkConfig",
    "MapMLGL",
    "MarkerMLGL",
    "PopupMLGL",
    "StyleMLGL",
    "CanvasSourceMLGL",
    "GeoJSONSourceMLGL",
    "ImageSourceMLGL",
    "RasterTileSourceMLGL",
    "RasterDEMTileSourceMLGL",
    "VectorTileSourceMLGL",
    "VideoSourceMLGL",
    "NavigationControMLGL",
    "GeolocateControlMLGL",
    "AttributionControlMLGL",
    "LogoControlMLGL",
    "ScaleControlMLGL",
    "FullscreenControlMLGL",
    "TerrainControMLGL",
    "BoxZoomHandlerMLGL",
    "ScrollZoomHandlerMLGL",
    "CooperativeGesturesHandlerMLGL",
    "KeyboardHandlerMLGL",
    "TwoFingersTouchPitchHandlerMLGL",
    "MapWheelEventMLGL",
    "MapTouchEventMLGL",
    "MapMouseEventMLGL",
    ...Object.keys(language),
    ...Object.keys(controls),
    ...Object.keys(converters),
    ...Object.keys(ColorRamp),
    ...Object.keys(helpers),
  ]),
);

describe("Module Exports", async () => {
  const exportedModule = await import("../src/index");

  console.info(
    "\x1b[1m%s\x1b[0m",
    `
  At present you will likely see an error above:
  
  \`Error: Failed to load url <blob-filepath> (resolved id: <blob-filepath>). Does the file exist?\`

  This is related to maplibre-gl and the way it handles CJS modules.
  It is caused by the call to \`enableRTL()\` in src/index.ts.

  There is a planned future work to remove this (the API it relies on
  is deprecated in Maplibre). But, for now, we must tolerate some
  noise in this test...
  `,
  );

  it("should match number of exptected exports with expected number of exports, logging any superfluous exports", () => {
    const actualExports = Object.keys(exportedModule);
    const superfluousExports = actualExports.filter((key) => !expectedExports.includes(key));
    if (superfluousExports.length > 0) {
      console.log("Unexpected exports found:", superfluousExports);
    }

    expect(Object.keys(exportedModule).length).toBe(expectedExports.length);
  });

  it("should export all expected values and log missing ones", () => {
    const missingExports = expectedExports.filter((key) => !(key in exportedModule));
    expect(missingExports).toEqual([]);
  });

  it("should have a valid version export", () => {
    expect(exportedModule.getVersion()).toBe(packagejson.version);
  });

  it("should match the snapshot of module exports", () => {
    // .sort() is used to ensure the order of keys is consistent and avoid brittle snapshots
    expect(Object.keys(exportedModule).sort()).toMatchSnapshot();
  });
});
