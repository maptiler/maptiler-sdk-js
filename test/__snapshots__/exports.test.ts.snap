// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Module Exports > should match the snapshot of module exports 1`] = `
[
  "AJAXError",
  "AttributionControl",
  "AttributionControlMLGL",
  "BoxZoomHandler",
  "BoxZoomHandlerMLGL",
  "CanvasSource",
  "CanvasSourceMLGL",
  "ColorRamp",
  "ColorRampCollection",
  "CooperativeGesturesHandler",
  "CooperativeGesturesHandlerMLGL",
  "DOMcreate",
  "DOMremove",
  "DoubleClickZoomHandler",
  "DragPanHandler",
  "DragRotateHandler",
  "EdgeInsets",
  "Evented",
  "FullscreenControl",
  "FullscreenControlMLGL",
  "GeoJSONSource",
  "GeoJSONSourceMLGL",
  "GeolocateControl",
  "GeolocateControlMLGL",
  "GeolocationType",
  "Hash",
  "ImageSource",
  "ImageSourceMLGL",
  "KeyboardHandler",
  "KeyboardHandlerMLGL",
  "Language",
  "LngLat",
  "LngLatBounds",
  "LogoControl",
  "LogoControlMLGL",
  "Map",
  "MapMLGL",
  "MapMouseEvent",
  "MapMouseEventMLGL",
  "MapStyle",
  "MapStyleVariant",
  "MapTouchEvent",
  "MapTouchEventMLGL",
  "MapWheelEvent",
  "MapWheelEventMLGL",
  "MaptilerGeolocateControl",
  "MaptilerLogoControl",
  "MaptilerNavigationControl",
  "MaptilerProjectionControl",
  "MaptilerTerrainControl",
  "Marker",
  "MarkerMLGL",
  "MercatorCoordinate",
  "NavigationControMLGL",
  "NavigationControl",
  "Point",
  "Popup",
  "PopupMLGL",
  "RasterDEMTileSource",
  "RasterDEMTileSourceMLGL",
  "RasterTileSource",
  "RasterTileSourceMLGL",
  "ReferenceMapStyle",
  "ScaleControl",
  "ScaleControlMLGL",
  "ScrollZoomHandler",
  "ScrollZoomHandlerMLGL",
  "SdkConfig",
  "ServiceError",
  "Style",
  "StyleMLGL",
  "TerrainControMLGL",
  "TerrainControl",
  "TwoFingersTouchPitchHandler",
  "TwoFingersTouchPitchHandlerMLGL",
  "TwoFingersTouchRotateHandler",
  "TwoFingersTouchZoomHandler",
  "TwoFingersTouchZoomRotateHandler",
  "VectorTileSource",
  "VectorTileSourceMLGL",
  "VideoSource",
  "VideoSourceMLGL",
  "addProtocol",
  "addSourceType",
  "areSameLanguages",
  "bufferToPixelDataBrowser",
  "circumferenceAtLatitude",
  "clearPrewarmedResources",
  "config",
  "configMLGL",
  "coordinates",
  "data",
  "displayWebGLContextLostWarning",
  "elevation",
  "expandMapStyle",
  "geocoding",
  "geolocation",
  "getAutoLanguage",
  "getBrowserLanguage",
  "getBufferToPixelDataParser",
  "getLanguageInfoFromCode",
  "getLanguageInfoFromFlag",
  "getLanguageInfoFromKey",
  "getMapLibreVersion",
  "getMaxParallelImageRequests",
  "getRTLTextPluginStatus",
  "getTileCache",
  "getVersion",
  "getWebGLSupportError",
  "getWorkerCount",
  "getWorkerUrl",
  "gpx",
  "gpxOrKml",
  "hasChildNodeWithName",
  "helpers",
  "importScriptInWorkers",
  "isLanguageInfo",
  "kml",
  "mapStylePresetList",
  "math",
  "misc",
  "prewarm",
  "removeProtocol",
  "setMaxParallelImageRequests",
  "setRTLTextPlugin",
  "setWorkerCount",
  "setWorkerUrl",
  "staticMaps",
  "str2xml",
  "styleToStyle",
  "toLanguageInfo",
  "xml2str",
]
`;
